File: 2017/22.py
from lib import disp
import math, copy, re, hashlib
import itertools as it

BURSTS1 = 10000
BURSTS2 = 10000000
DIRS = "NESW"
MOVES = {
	"N": (0, -1),
	"E": (1, 0),
	"S": (0, 1),
	"W": (-1, 0)
}

"""
# - infected
. - clean
~ - weakened
! - flagged
"""

NEWD = {
	'#': (1, '!'),
	'.': (-1, '~'),
	'~': (0, '#'),
	'!': (2, '.')
}

def get_dir(d, curr):
	i = DIRS.index(d)
	i += 1 if curr == '#' else -1
	return DIRS[i%4]

def get_pos(p, d):
	dx, dy = MOVES[d]
	return [p[0] + dx, p[1] + dy]


def get_ext_dir(d, curr):
	di = NEWD[curr][0]
	i = DIRS.index(d) + di
	return DIRS[i%4]

def part_1(data):
	l = len(data)
	M = 200
	GRID = [['.' for i in range(M*l)] for j in range(M*l)]
	for i in range(l):
		for j in range(l):
			GRID[l*M//2 + i][l*M//2 + j] = data[i][j]
	pos = [l // 2 + l*M//2, l // 2 + l*M//2]
	d = 'N'
	i = 0
	inf = 0
	
	while i < BURSTS1:
		[x, y] = pos
		# print(x, y)
		prev = GRID[y][x]
		if GRID[y][x] == '.':
			GRID[y][x] = '#'
			inf += 1
		else:
			GRID[y][x] = '.'

		d = get_dir(d, prev)
		pos = get_pos(pos, d)
		i += 1

	print(inf)
	print('END OF PART1')
	return

def part_2(data):
	l = len(data)
	M = 200
	GRID = [['.' for i in range(M*l)] for j in range(M*l)]
	for i in range(l):
		for j in range(l):
			GRID[l*M//2 + i][l*M//2 + j] = data[i][j]
	pos = [l // 2 + l*M//2, l // 2 + l*M//2]
	d = 'N'
	i = 0
	inf = 0
	
	while i < BURSTS2:
		[x, y] = pos
		# print(x, y)
		prev = GRID[y][x]
		GRID[y][x] = NEWD[prev][1]
		if GRID[y][x] == '#':
			inf += 1

		d = get_ext_dir(d, prev)
		pos = get_pos(pos, d)
		i += 1
	print(inf)
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('22_input') as f:
		data = f.read()
		data = data.split('\n')
		# data = list(map(int, data.split()))


	part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2017/22.js

const fs = require('fs');
const { exec } = require("child_process");
const { ALPHA_L, ALPHA_U, D4, D6, D8, MOD } = require("../lib");
const { joseph, findCycle } = require("../lib");
const { cache } = require("../lib");
const { bin, float, hex, int, num, oct } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { copy,	entries, in_, inBB, keys, sort, values } = require("../lib");
const { Counter, DD, empty, FastQueue, PriorityQueue, Queue, set, Stack } = require("../lib");
const { areaInt, circumference, manDist, shoelace } = require("../lib");
const { ord, chr, count, debug, disp, disp3, crt, gcd, lcm, modPow, modPowBig, modInv, mod, prod, prodBig, randint, sum, sumBig, transpose } = require("../lib");
const { digits, ints,	floats,	singles,	words,	lines,	table,	groups,	getGroups,	groupsWith,	parse,	parseLine, scanf } = require ('../lib');
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;
const { getExecStr } = require("../lib/post");
const { combinations, combinations_with_replacement, next_permutation, product } = require("../lib");

const right = (dy, dx) => [dx, -dy];
const left = (dy, dx) => [-dx, dy];

function part1(data) {

	let res = 0;
	const G = {};
	data = table(data);
	drange(data.length)((i,j) => {
		G[[i, j]] = data[i][j];
	})
	let x = int(data.length / 2);
	let y = int(data.length / 2);
	let dy = -1, dx = 0;
	range(10_000)(_ => {
		if (!G[[y,x]]) G[[y,x]] = '.';
		if (G[[y,x]] === '#') {
			[dy, dx] = right(dy, dx);
			G[[y,x]] = '.';
		} else {
			[dy, dx] = left(dy, dx);
			G[[y,x]] = '#';
			++res;
		}
		[y, x] = [y + dy, x + dx];
	});

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART1");
	return;
}

const change = {
	'#': { // infected
		nxt: '$',
		dir: right
	},
	'$': { // flagged
		nxt: '.',
		dir: (dy, dx) => [-dy, -dx]
	},
	'.': { // clean
		nxt: '!',
		dir: left
	},
	'!': { // weakened
		nxt: '#',
		dir: (dy, dx) => [dy, dx]
	}
}

function part2(data) {

	let res = 0;
	const G = {};
	data = table(data);
	drange(data.length)((i,j) => {
		G[[i, j]] = data[i][j];
	})
	let x = int(data.length / 2);
	let y = int(data.length / 2);
	let dy = -1, dx = 0;
	range(10_000_000)(_ => {
		if (!G[[y,x]]) G[[y,x]] = '.';
		[dy, dx] = change[ G[[y,x]] ].dir(dy, dx);
		G[[y, x]] = change[ G[[y,x]] ].nxt;
		if (G[[y, x]] === '#') ++res;
		[y, x] = [y + dy, x + dx];
	});

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("22_input").toString("utf-8");

	part1(data);
	part2(data);
	process.exit(0);
}

main();
